userguide/cpp.md

```diff
@@ -0,0 +1,198 @@
+// LD_LIBRARY_PATH=../turicreate/release/src/unity ./a.out
+
+// standard C++ includes
+#include <iostream>
+
+// Turi Create includes
+#include <logger/assertions.hpp>
+#include <logger/logger.hpp>
+#include <unity/lib/unity_global.hpp>
+#include <unity/server/registration.hpp>
+#include <unity/toolkits/supervised_learning/boosted_trees.hpp>
+
+int main(int argc, char **argv) {
+  // set up registry of named functions and models
+  // used for rehydrating saved models
+  auto fn_reg = turi::toolkit_function_registry();
+  turi::register_functions(fn_reg);
+  auto class_reg = turi::toolkit_class_registry();
+  turi::register_models(class_reg);
+
+  // create a global context
+  auto global = turi::unity_global(&fn_reg, &class_reg);
+
+  // load the model from disk
+  auto loaded_model = global.load_model("mushroom.tcmodel");
+  std::string name = turi::variant_get_value<turi::flexible_type>(loaded_model["model_name"]);
+  std::cout << "Loaded model: " << name << std::endl;
+  auto model = std::dynamic_pointer_cast<turi::supervised::xgboost::boosted_trees_classifier>(
+    turi::variant_get_value<std::shared_ptr<turi::model_base>>(loaded_model["model"])
+  );
+
+  // set up input
+  // (note: can do batch prediction by adding more than one row to rows)
+  std::vector<turi::flexible_type> rows;
+  turi::flex_dict row;
+  row.push_back(std::make_pair("bruises?", "t"));
+  row.push_back(std::make_pair("cap-color", "n"));
+  row.push_back(std::make_pair("cap-shape", "x"));
+  row.push_back(std::make_pair("cap-surface", "s"));
+  row.push_back(std::make_pair("gill-attachment", "f"));
+  row.push_back(std::make_pair("gill-color", "k"));
+  row.push_back(std::make_pair("gill-size", "n"));
+  row.push_back(std::make_pair("gill-spacing", "c"));
+  row.push_back(std::make_pair("habitat", "u"));
+  row.push_back(std::make_pair("odor", "p"));
+  row.push_back(std::make_pair("population", "s"));
+  row.push_back(std::make_pair("ring-number", "o"));
+  row.push_back(std::make_pair("ring-type", "p"));
+  row.push_back(std::make_pair("spore-print-color", "k"));
+  row.push_back(std::make_pair("stalk-color-above-ring", "w"));
+  row.push_back(std::make_pair("stalk-color-below-ring", "w"));
+  row.push_back(std::make_pair("stalk-root", "e"));
+  row.push_back(std::make_pair("stalk-shape", "e"));
+  row.push_back(std::make_pair("stalk-surface-above-ring", "s"));
+  row.push_back(std::make_pair("stalk-surface-below-ring", "s"));
+  row.push_back(std::make_pair("veil-color", "w"));
+  row.push_back(std::make_pair("veil-type", "p"));
+  rows.push_back(row);
+
+  // make prediction
+  // (returns a gl_sframe)
+  turi::gl_sframe result = model->fast_predict_topk(
+      // input data - 1 row per prediction
+      rows,
+
+      // output type - see https://github.com/apple/turicreate/blob/3490286b27ff5d79cb90d09fe026d5671ce990c7/src/unity/toolkits/supervised_learning/supervised_learning.hpp#L47
```

Comment:
Links to line numbers can easily get out of date. I think if people are going to be linking against us they need to be responsible for figure some of this stuff out themselves. I would just remove this link and the one on line 165.

Reply:
This link goes to a specific commit hash, so the line number will stay consistent.